onnxruntime/core/session/inference_session.cc

```diff
@@ -1930,6 +1931,44 @@
 #ifdef DEBUG_NODE_INPUTS_OUTPUTS
       session_state_->IncrementGraphExecutionCounter();
 #endif
+
+      struct ThreadPoolSpinningSwitch {
+        using PS = onnxruntime::concurrency::ThreadPool::ParallelSection;
+        concurrency::ThreadPool* intra_tp_;
+        concurrency::ThreadPool* inter_tp_;
+        std::atomic_int32_t& counter_ref_;
+        // Use this to jump start threads and amortize the costs
+        // of initialization between the kernels
+        // note this prevents using explicit PS in the nodes
+        // or an additional PS for inter op thread-pool
+        std::optional<PS> ps_intra_;
+        ThreadPoolSpinningSwitch(concurrency::ThreadPool* intra_tp,
+                                 concurrency::ThreadPool* inter_tp,
+                                 std::atomic_int32_t& ref) noexcept
+            : intra_tp_(intra_tp), inter_tp_(inter_tp), counter_ref_(ref) {
+          if (counter_ref_.fetch_add(1, std::memory_order_relaxed) == 0) {
+            if (intra_tp_) intra_tp_->EnableSpinning();
+            if (inter_tp_) inter_tp_->EnableSpinning();
+          }
+          if (intra_tp_) {
+            ps_intra_.emplace(intra_tp_);
+          }
+        }
+        ~ThreadPoolSpinningSwitch() {
+          ps_intra_.reset();
+          if (1 == counter_ref_.fetch_sub(1, std::memory_order_acq_rel)) {
+            if (intra_tp_) intra_tp_->DisableSpinning();
+            if (inter_tp_) inter_tp_->DisableSpinning();
+          }
+        }
+      };
+
+      std::optional<ThreadPoolSpinningSwitch> tp_starter;
+      if (force_spinning_stop_between_runs_) {
+        concurrency::ThreadPool* intra_tp_ = (use_per_session_threads_) ? thread_pool_.get() : intra_op_thread_pool_from_env_;
+        concurrency::ThreadPool* inter_tp = (use_per_session_threads_) ? inter_op_thread_pool_.get() : inter_op_thread_pool_from_env_;
+        tp_starter.emplace(intra_tp_, inter_tp, invocation_refcounter_);
+      }
```

Comment:
What if custom slightly altered their scenario - say they running 2/3/4 inferences at a time, they want to stop spinning and restart in between each 2/3/4 calls?

Reply:
If we place two API calls:

session::stop_spinning(...)
session::start_spinning(...)

would it be a bit more versatile?

Reply:
Yes, running multiple instance at the time is supported by this code.
Calling start/stop spinning calls for a new API. That would fall under TP API. I would prefer not to add new API unless is required. It also should work properly with the PS, which is an implementation detail.

Reply:
I mean sequential calling of Run() in a single main thread for 2/3/4 times in a row, not concurrently.

Reply:
It will start and stop them between the runs and it necessary. This is exact scenario that we were asked for.
There might be continuous calls (I have verified that it does not degrade perf) and there might be periods of infrequent calls. Spinning is helpful when Run() is in progress.